File: Language_Essentials/python/q7_two_pointer.py
# Two-Pointer Technique - TCS NQT

def has_pair_with_sum(arr, target):
    left = 0
    right = len(arr) - 1
    
    while left < right:
        current_sum = arr[left] + arr[right]
        
        if current_sum == target:
            print(f"Pair found: {arr[left]} + {arr[right]} = {target}")
            return True
        elif current_sum < target:
            left += 1  # Need larger sum
        else:
            right -= 1 # Need smaller sum
            
    print(f"No pair found for target {target}")
    return False

def main():
    arr = [2, 4, 6, 8, 10, 12, 14]
    target = 18
    
    print("Array:", arr)
    print("Target:", target)
    
    has_pair_with_sum(arr, target)

if __name__ == "__main__":
    main()

File: Language_Essentials/java/Q7TwoPointer.java
// Two-Pointer Technique - TCS NQT
import java.util.Arrays;

public class Q7TwoPointer {
    
    // Find pair with given sum in a sorted array
    public static boolean hasPairWithSum(int[] arr, int target) {
        int left = 0;
        int right = arr.length - 1;
        
        while (left < right) {
            int currentSum = arr[left] + arr[right];
            
            if (currentSum == target) {
                System.out.println("Pair found: " + arr[left] + " + " + arr[right] + " = " + target);
                return true;
            } else if (currentSum < target) {
                left++; // Need larger sum
            } else {
                right--; // Need smaller sum
            }
        }
        System.out.println("No pair found for target " + target);
        return false;
    }

    public static void main(String[] args) {
        int[] arr = {2, 4, 6, 8, 10, 12, 14};
        int target = 18;
        
        System.out.println("Array: " + Arrays.toString(arr));
        System.out.println("Target: " + target);
        
        hasPairWithSum(arr, target);
    }
}
